plugin/core/promise.py

```diff
@@ -0,0 +1,194 @@
+from .typing import Any, List, Optional
+import functools
+import threading
+
+
+class PromiseError(Exception):
+    """A failed Promise is to be resolved with this PromiseError.
+
+    Normal errors and exceptions can't be caught by the application due to
+    multi-threading. Therefore a function should return PromiseError to indicate
+    a failure to be handled by `.then()`
+    """
+    pass
```

Comment:
It's not used now but if there would be a case where we would want to indicate that promise is returning an error then it would be useful. Or another option would be to implement `catch` and `Promise.reject()` but that's relatively large amount of work.

Reply:
Removed for now.